### ISSUE TITLE
fix(versions): syncpack crashes on pnpm alias override

### ISSUE THREAD

**Issue by ari-becker** [JamieMason/syncpack#165]
## Description

`pnpm exec syncpack --version` is `11.2.1`.

Run `pnpm exec synpack list --types=optional`.

Output:

```text
/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/runtime.js:69
    throw fiberFailure(result.i0);
    ^

Error [TypeError]: Cannot read properties of undefined (reading 'read')
TypeError: Cannot read properties of undefined (reading 'read')
    at /Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-package-json-files/package-json-file.js:22:43
    at Array.forEach (<anonymous>)
    at PackageJsonFile.getInstances (/Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-package-json-files/package-json-file.js:21:22)
    at EffectPrimitive.i1 (/Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-version-groups/index.js:134:52)
    at Object.OnSuccess (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:88:17)
    at FiberRuntime.Sync (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:859:38)
    at /Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:1066:29
    at Const.onRun (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/supervisor.js:167:12)
    at FiberRuntime.runLoop (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:1064:32)
    at FiberRuntime.evaluateEffect (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:675:29) {
  toString: [Function (anonymous)],
  [Symbol(@effect/io/Runtime/FiberFailure)]: Symbol(@effect/io/Runtime/FiberFailure),
  [Symbol(@effect/io/Runtime/FiberFailure/Cause)]: {
    _tag: 'Die',
    defect: TypeError: Cannot read properties of undefined (reading 'read')
        at /Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-package-json-files/package-json-file.js:22:43
        at Array.forEach (<anonymous>)
        at PackageJsonFile.getInstances (/Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-package-json-files/package-json-file.js:21:22)
        at EffectPrimitive.i1 (/Users/me/path/to/repo/node_modules/.pnpm/syncpack@11.2.1/node_modules/syncpack/dist/get-version-groups/index.js:134:52)
        at Object.OnSuccess (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:88:17)
        at FiberRuntime.Sync (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:859:38)
        at /Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:1066:29
        at Const.onRun (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/supervisor.js:167:12)
        at FiberRuntime.runLoop (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:1064:32)
        at FiberRuntime.evaluateEffect (/Users/me/path/to/repo/node_modules/.pnpm/@effect+io@0.38.0_@effect+data@0.17.1/node_modules/@effect/io/internal/fiberRuntime.js:675:29)
  },
  [Symbol(nodejs.util.inspect.custom)]: [Function (anonymous)]
}

Node.js v18.17.1

```

## Suggested Solution

Command should run normally

## Help Needed

N/A

**Comment by JamieMason** [comment]
👋 there is no dependency type called optional

**Comment by ari-becker** [comment]
@JamieMason I added `optional` to try to make it easier to help debug this.
I get the same error when I run `pnpm exec syncpack list`, but I don't get an error when I run `pnpm exec syncpack list --types=prod,dev,peer`.

My `pnpm-lock.yaml` includes `optionalDependencies`, including e.g.:

![image](https://github.com/JamieMason/syncpack/assets/1388505/2bc8366e-aa8f-464a-a3b3-f490524a1190)

**Comment by JamieMason** [comment]
Can you share a reproduction please

**Comment by ari-becker** [comment]
> Can you share a reproduction please

OK @JamieMason I found you a reproduction: https://github.com/ari-becker/synpack-165-repro . sorry it's not quite minimal (I added stuff and added stuff until I found the trigger) but it's pretty minimal.

The issue is:

`/package.json`:

```json
{
  "pnpm": {
    "overrides": {
      "react-icons": "npm:@lorenzobloedow/react-icons"
    }
  }
}
```

`/packages/foo/package.json`:

```json
{
  "dependencies": {
    "react-icons": "^4.10.1"
  }
}
```

This uses `pnpm`'s ability to replace a dependency with a fork: https://pnpm.io/package_json#pnpmoverrides . In this particular case, to use a fork for `react-icons` to solve an issue in the core `react-icons` package not including `"type":"module"`: https://github.com/react-icons/react-icons/issues/717#issuecomment-1648271176

**Comment by JamieMason** [comment]
Thanks a lot Ari, I'll take a look.

**Comment by JamieMason** [comment]
Just from looking at the source on GitHub, I think the problem as you mentioned is that syncpack 11.x does not understand the non-semver version of `npm:@lorenzobloedow/react-icons`.

There is a [new version coming](https://github.com/JamieMason/syncpack/compare/main...dev?diff=unified#diff) in maybe 3 weeks or so which has better support for this and other non-semver version formats. I'll add a test case for your scenario to be sure.

**Comment by ari-becker** [comment]
> Just from looking at the source on GitHub, I think the problem as you mentioned is that syncpack 11.x does not understand the non-semver version of `npm:@lorenzobloedow/react-icons`.
> 
> There is a [new version coming](https://github.com/JamieMason/syncpack/compare/main...dev?diff=unified#diff) in maybe 3 weeks or so which has better support for this and other non-semver version formats. I'll add a test case for your scenario to be sure.

Fantastic, thank you!

For the moment, finding the reproduction helped me find a workaround by pinning the fork to a specific version - the following solves the issue for me:

```json
{
  "pnpm": {
    "overrides": {
      "react-icons": "npm:@lorenzobloedow/react-icons@1.0.0",
    }
  }
}
```

**Comment by JamieMason** [comment]
Released [12.0.0-alpha.0](https://github.com/JamieMason/syncpack/releases/tag/12.0.0-alpha.0) which should handle this better, but I'll take a look at it properly when I get time.

**Comment by JamieMason** [comment]
Closing as I think v12 handles this as expected, it reports on which versions it does not understand rather than throwing.

<details>
  <summary>list-mismatches output</summary>

```
= Default Version Group ========================================================
✘ react-icons ^4.10.1 → npm:@lorenzobloedow/react-icons packages/foo/package.json > dependencies [HighestSemverMismatch]
✘ typescript ^5.1.6 → ??? package.json > devDependencies [UnsupportedMismatch]
  use syncpack prompt to fix this manually
✘ typescript $typescript → ??? package.json > pnpm.overrides [UnsupportedMismatch]
  use syncpack prompt to fix this manually
✘ typescript * → ??? packages/foo/package.json > devDependencies [UnsupportedMismatch]
  use syncpack prompt to fix this manually
    27 ✓ already valid
     1 ✓ can be auto-fixed
     3 ! can be fixed manually using syncpack prompt
```

</details>

<details>
  <summary>list output</summary>

```
= Default Version Group ========================================================
     1x @storybook/addon-essentials: ^7.4.6
     1x @storybook/addon-interactions: ^7.4.6
     1x @storybook/addon-links: ^7.4.6
     1x @storybook/addon-onboarding: ^1.0.8
     1x @storybook/blocks: ^7.4.6
     1x @storybook/nextjs: ^7.4.6
     1x @storybook/react: ^7.4.6
     1x @storybook/testing-library: ^0.2.2
     1x @t3-oss/env-nextjs: ^0.7.0
     1x @types/eslint: ^8.44.2
     1x @types/node: ^18.16.0
     1x @types/react: ^18.2.20
     1x @types/react-dom: ^18.2.7
     1x @typescript-eslint/eslint-plugin: ^6.3.0
     1x @typescript-eslint/parser: ^6.3.0
     1x eslint: ^8.47.0
     1x eslint-config-next: ^13.5.4
     1x eslint-plugin-storybook: ^0.6.15
     1x foo: 0.1.0
     1x next: ^13.5.4
     1x react: 18.2.0
     1x react-dom: 18.2.0
     2x react-icons:
        ✘ ^4.10.1 → npm:@lorenzobloedow/react-icons [HighestSemverMismatch]
     1x storybook: ^7.4.6
     1x syncpack: ^11.2.1
     1x synpack-165-repro: 0.1.0
     3x typescript:
        ✘ ^5.1.6 → [UnsupportedMismatch]
        ✘ $typescript → [UnsupportedMismatch]
        ✘ * → [UnsupportedMismatch]
     1x zod: ^3.22.4

  What next?
- syncpack list-mismatches to see more detail about mismatching versions
- syncpack fix-mismatches to fix version mismatches automatically
- syncpack format to sort and prettify your package.json files
- syncpack update to choose updates from the npm registry
- syncpack --help for everything else
```

</details>
